std/algorithm/comparison.d

```diff
@@ -1565,3 +1571,407 @@ unittest
+    auto r6 = new ReferenceInputRange!int([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
+    assert(isSameLength(r5, r6));
+
+    auto r7 = new ReferenceInputRange!int([1, 2]);
+    auto r8 = new ReferenceInputRange!int([4, 5, 6]);
+    assert(!isSameLength(r7, r8));
+
+    auto r9 = new ReferenceInputRange!int([1, 2, 3, 4, 5, 6, 7, 8]);
+    DummyRange!(ReturnBy.Reference, Length.Yes, RangeType.Input) r10;
+    assert(!isSameLength(r9, r10));
+
+    DummyRange!(ReturnBy.Reference, Length.Yes, RangeType.Input) r11;
+    auto r12 = new ReferenceInputRange!int([1, 2, 3, 4, 5, 6, 7, 8]);
+    assert(!isSameLength(r11, r12));
+}
+
+alias AllocateGC = Flag!"AllocateGC";
+
+/**
+Returns $(D true) if all of the elements in $(D r1) appear the same number of times
+in $(D r2). Otherwise, returns $(D false).
+
+This function can allocate if the $(D AllocateGC.yes) flag is passed. This has
+the benefit of have better complexity than the $(D AllocateGC.no) option. However,
+this option is only available for ranges who's equality can be determined via each
+element's $(D toHash) method. If customized equality is needed, then the $(D pred)
+template parameter can be passed, and the function will automatically switch to
+the non-allocating algorithm. See $(D binaryFun) for more details on how to define
+$(D pred).
+
+Non allocating forward range option: $(BIGOH n^2)
+Non allocating forward range option with custom $(D pred): $(BIGOH n^2)
+Allocating forward range option: amortized $(BIGOH r1.length) + $(BIGOH r2.length)
+
+Params:
+    pred = an optional parameter to change how equality is defined
+    allocate_gc = A $(D std.typecons.Flag!"AllocateGC") instance
+    r1 = A finite forward range
+    r2 = A finite forward range
+
+Returns:
+    $(D bool)
+*/
+
+bool isPermutation(AllocateGC allocate_gc, Range1, Range2)
+(Range1 r1, Range2 r2)
+    if (allocate_gc == AllocateGC.yes &&
+        isForwardRange!Range1 &&
+        isForwardRange!Range2 &&
+        !isInfinite!Range1 &&
+        !isInfinite!Range2)
+{
+    alias E1 = Unqual!(ElementType!Range1);
+    alias E2 = Unqual!(ElementType!Range2);
+
+    if(!isSameLength(r1.save, r2.save))
+    {
+        return false;
+    }
+
+    // Skip the elements at the beginning where r1.front == r2.front,
+    // they are in the same order and don't need to be counted.
+    while (!r1.empty && !r2.empty)
+    {
+        if (r1.front == r2.front)
+        {
+            r1.popFront;
+            r2.popFront;
+        }
+        else
+        {
+            break;
+        }
+    }
+
+    if (r1.empty && r2.empty)
+    {
+        return true;
+    }
+
+    static if (allocate_gc == AllocateGC.yes)
+    {
+        int[CommonType!(E1, E2)] counts;
+
+        foreach (item; r1)
+        {
+            ++counts[item];
+        }
+
+        foreach (item; r2)
+        {
+            if (--counts[item] < 0)
+            {
+                return false;
+            }
+        }
+
+        return true;
+    }
+    else
```

Comment:
Can get rid of this branch. assert(0) instead. This instantiation shouldn't happen unless allocate_gc is no.